### PR TITLE
del logits=(bs, seq_len, vocab_size) to save 3.9G memory

### PR DESCRIPTION
**Pull request by weifengpy**
logits=(bs, seq_len, vocab_size). call `del logits` to free it before backward

Screenshot 2024-06-12 at 11 10 36 AM

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -351,6 +351,9 @@ def loss_fn(pred, labels):
                 with loss_parallel_ctx():
                     pred = model(input_ids)
                     loss = loss_fn(pred, labels)
+                    # pred.shape=(bs, seq_len, vocab_size)
+                    # need to free to before bwd to avoid peaking memory
+                    del pred
```
Comment (wanchaol) [inline train.py:356]: hmmm I wonder could we really del or free this? iiuc the `pred` is part of the autograd graph that needs to be there for backward computation?

Comment (weifengpy) [reply train.py:356]: I believe `pred` is passed into `loss_fn` and `loss` tracks all the autograd ? We did the same trick in xlformer https://github.com/fairinternal/xlformers/blame/001955c08b7a2c5e008d5c503af985f8c80549f2/train.py#L777

or do you prefer some test to verify numerics ?

Comment (wanchaol) [reply train.py:356]: after some thoughts I think this actually make sense to me!

**Review by wanchaol — APPROVED** [review]

**Comment by bdhirsh** [comment]
nice!

File: train.py (hunk 1 above)
Comment (weifengpy) [reply train.py:356]: the linter installation in CI seems to be broken, since it appears in other PRs as well. For example, this PR changes README https://github.com/pytorch/torchtitan/pull/385 and triggers the same error 

thus landing this PR for now
